### PR TITLE
MNG-7395 - Support system-properties in extension.xml

### PR DESCRIPTION
**Pull request by laeubi**
We at [Tycho](https://github.com/eclipse/tycho) have a combination of a core-extension and a build-plugin. What always is a source of confusion, that one is able to define e.g. a `<tycho-version>` property that could be overridden on the commandline using -D but no such thing exits for the `.mvn/extensions.xml`

This makes it also hard to make some regression-testing e.g. to check fast if a certain problem occurs also in the previous version as one always has to take in mind that the `.mvn/extensions.xml` must be edited by hand.

For that reason it would help if at least basic interpolation using the "maven-properties" would be possible.

This PR adds support for having basic interpolation in `.mvn/extensions.xml` to either:

1. Specify a (default) version in `.mvn/jvm.config`
2. Specify a version on the commandline with `mvn -D=some.version=1.2.3`

FYI @mickaelistria @akurtakov 

Following this checklist to help us incorporate your
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed
       for the change (usually before you start working on it).  Trivial changes like typos do not
       require a JIRA issue. Your pull request should address just this issue, without
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] SUMMARY`, where you replace `MNG-XXX`
       and `SUMMARY` with the appropriate JIRA issue. Best practice is to use the JIRA issue
       title in the pull request title and in the first line of the commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will
       be performed on your pull request automatically.
 - [ ] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

**Review by michael-o — CHANGES_REQUESTED** [review]
* Interpolation happen the same way it happens for other models
* If we are going to allow default values, the approach would be a subset of this: https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_06

**Comment by laeubi** [comment]
> * If we are going to allow default values, the approach would be a subset of this

So you suggest something like:

`${tycho-version:=2.6.0}`

to have default values?

**Comment by michael-o** [comment]
> > * If we are going to allow default values, the approach would be a subset of this
> 
> So you suggest something like:
> 
> `${tycho-version:=2.6.0}`
> 
> to have default values?

Rather `parameter:-word`

**Comment by laeubi** [comment]
> I don't see the need for that since in the pom you can set in properties `<myextension.version>1.2.3</>` and override it with `-Dmyextension.version` in maven's spirit.

Please take a look at the code and/or [referenced JIRA ](https://issues.apache.org/jira/browse/MNG-7395), at the time the extensions are fetched/created there is no properties, no session, even no "maven".
If you still think it is possible a small example would be useful (disclaimer: I have already tried this out and got only an warning from maven where it complains that `${myextension.version}` could not be found as a version in any repository) you can use the one I have [attached here](https://issues.apache.org/jira/secure/attachment/13039302/testit.zip) as a starting point.

> Anything else just creates something inconsistent without new real features so better to ensure the "on the fly plugin" is processed as any model instead of making it specific (and lacking env support for ex), no?

Why do you think it is inconsistent? It is what most maven developers would expect and the first thing you have to learn when using core-extensions  with `.mvn/extensions.xml` that they are simply dump and missing all features beside plain text parsing. So I'd like to have *at laest* support for java properties, if one feels env-variables are useful or whatever else that is avaiable at this very early stage, why not, but at the end of the day I'd better have support for "only" system properties instead of no support at all because there are so many other thing one *might* could need and essentially the feature never happens.

**Comment by laeubi** [comment]
@michael-o I now added the interrogator similar to that method you mentioned
@rmannibucau It now also support env-variables.

**Comment by rmannibucau** [comment]
@laeubi I read the jira and this is why i commented this way. Does `.mvn/extensions.xml` relates to a project? Yes. So it can use the standard maven filtering. The only limitation will be that it is preprocessed, ie you can't use extension enrichment there - by design - but you can use project model so properties and you can use system properties and env variable interpolation as in any part of a maven filtering so no need to invent a new one.

>  It is what most maven developers would expect

I think most maven users will expect filtering to behave as maven filtering, not with a spring/unix/... alternative.

So overall it is very good to bring filtering while it is aligned on maven one and not reimplemented+reinvented IMHO.

**Comment by laeubi** [comment]
> Does `.mvn/extensions.xml` relates to a project?

Not that I'm aware of. At that point of time there are no projects and thus no properties. It behaves as if you put the desired extension into `$MVN_HOME/lib/ext`

**Comment by rmannibucau** [comment]
@laeubi `.mvn` only makes sense in the context of a project so it does by design so code must be updated to get the raw model and use standard maven filtering on it IMHO. Don't get me wrong, I understand it is not the current implementation but it is the one we should aim for if we want to support filtering in maven ecosystem I think, others create inconsistencies for users as expained.

**Comment by laeubi** [comment]
> only makes sense in the context of a project

Don't think so a core-extension can simply extend everything in maven, even what "project" means ... Just take a look at the [polyglot-extension](https://github.com/takari/polyglot-maven) that allows alternative markup languages instead of pom.xml

Now assume I define my project in terms of json.properties, how should this "project" ever become materialized if the extension required to read it is required to first read the project to get the version property to know what version of the extension should be loaded ... I think its clear that this will never be resolvable...

**Comment by rmannibucau** [comment]
@laeubi well there are two comments to this point:

1. it does not justify to use a new way to interpolate values, once again I think we should tend to stick to our own convention and usage and not multiply syntaxes for the same purpose
2. the extensions are loaded from the clirequest after the env/system props so still the standard interpolation can be used even if part of the project is not available (worse case you just need to adapt `PluginParameterExpressionEvaluator` to ignore the session related variables). I still think trying to read the raw root pom in a fast pass to extract the properties if present since it will be available in most cases and it would enable to use defaults there but not a blocker to me to not have defaults there since `.mvn/jvm.config` enables to set them if needed in all cases.

**Comment by laeubi** [comment]
> 1\. it does not justify to use a new way to interpolate values

As mentioned above I now use what maven uses (or more, whats used on other places), the only extension is the default value, and I think this is valid here as

1. you don't need to use it if you don't like it
2. there is no other way to express defaults

> 2\. the extensions are loaded from the clirequest after the env/system props so still the standard interpolation

This is already done, see my latest changes and even env variables are possible now

> 2\. I still think trying to read the raw root pom in a fast pass to extract the properties

It is only if you don't consider parents and all that stuff

> 2\. will be available in most cases and it would enable to use defaults there but not a blocker to me to not have defaults there since `.mvn/jvm.config` enables to set them if needed in all cases.

The only reference to this seems the wrapper scripts, so is this something specific to the wrapper scripts or why is it not handled by the cli itself?
It also seems to indicate to be reserved for JVM parameter (even though properties are parameters it looks a bit strange) so won't these not be defaults but overrides what is passed on the commandline?

**Comment by rmannibucau** [comment]
> I now use what maven uses

About the env, I think the env vars are added to properties as `env.<name>` so `EnvarBasedValueSource` is not needed, isnt it? Otherwise ok for me for this part.

> you don't need to use it if you don't like it

It is not about me but what you expose to users. I'll present it differently: if you want to support defaults there you must support them everywhere to ensure all maven filtering is consistent. AFAIK maven community always had been against a mini language for that so I think it is sane to stick to what there is, in particular since inline defaults do not enable any use case for end users as explained.

> there is no other way to express defaults

I explained you it is not true in my previous comment so this part is solved.

> It is only if you don't consider parents and all that stuff

Right, this is what I had in mind but no issue to drop that since it is not really needed, was just convenient.

> The only reference to this seems the wrapper scripts, so is this something specific to the wrapper scripts or why is it not handled by the cli itself?

because it must be set on the JVM before it is launched (think you can configure JUL `LogManager` there for example).

> It also seems to indicate to be reserved for JVM parameter (even though properties are parameters it looks a bit strange) so won't these not be defaults but overrides what is passed on the commandline?

You will set `-Dmyprop=mydefault` on the JVM and maven will process the CLI args and override it normally so think it is ok.

**Comment by michael-o** [comment]
Please not that system properties != Maven properties. In Maven 4 the hack of property promotion will likely go away.

**Comment by laeubi** [comment]
> About the env, I think the env vars are added to properties as `env.<name>` so `EnvarBasedValueSource` is not needed, isnt it? Otherwise ok for me for this part.

EnvarBasedValueSource seems to handel the env. prefix and as it is used at other places it seems fine...

> It is not about me but what you expose to users. I'll present it differently: if you want to support defaults there you must support them everywhere

It could be added to PropertyBasedValueSource if that's useful or we can drop this if that's the only concern.

> You will set `-Dmyprop=mydefault` on the JVM and maven will process the CLI args and override it normally

all right, then this seems a bit like a hidden-feature (to me) so probably we don't need defaults in a strict way, I still think such a syntax would be nice (here) to have everything at one place. But as mentioned before I'm not insist on having default values, it was just an easy to archive option I though bringing some value here.

**Comment by rmannibucau** [comment]
Just meant `EnvarBasedValueSource` being dropped should keep it functional since system properties contain env vars so can avoid the add+lack of exception handling. So +1 to drop and just harness it with a test.

@laeubi don't get me wrong, I'm not strictly against having defaults (and actually a bit more ;)) but I'm against not having it everywhere. So both options are valid for me: drop defaults OR support them everywhere (but this last one will likely break the ecosystem so should be considered with caution since a lot of pom readers will not be aware of it).

BTW thanks a lot for this feature, was maybe not obvious but I think it is a very good one!

**Comment by laeubi** [comment]
I'll remove the defaults (for now) and just propose a more general approach in a separate PR. I'm really keen to get this in as it bugged me too many times :-D

**Comment by laeubi** [comment]
I have suashed the commits and using only "standard-maven" now, so I think this is ready for review.

**Review by rmannibucau — APPROVED** [review]
Looks like some indentation fix should be done but once fixed PR looks good to me.

**Comment by laeubi** [comment]
> Looks like some indentation fix should be done but once fixed PR looks good to me.

I have formatted with the "maven eclipse  profile for eclipse" now and hope everything is fine now, how can we further proceed here? Let me know if there are any issues left. As far as I know I have signed the Individual Contributor License Agreement a while back, how can I check the status here if it is still valid (if necessary)?

In any case I hereby declare this contribution to be licenced under the Apache License Version 2.0, January 2004

File: maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java [hunk 1]
```diff
@@ -127,30 +134,39 @@ private CoreExtensionEntry createExtension( CoreExtension extension, List<Artifa
...
+        catch ( PluginResolutionException | InterpolationException e )
         {
             throw new ExtensionResolutionException( extension, e.getCause() );
         }
     }
 
+    private static Interpolator createInterpolatorFunction( MavenExecutionRequest request )
+    {
+        RegexBasedInterpolator interpolator = new RegexBasedInterpolator();
+        interpolator.addValueSource( new PropertiesBasedValueSource( request.getUserProperties() ) );
```
Comment (michael-o) [inline maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java:167]: Please use for consistency reasons `MapBasedValueSource`

File: maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java [hunk 2]
```diff
@@ -127,30 +134,39 @@ private CoreExtensionEntry createExtension( CoreExtension extension, List<Artifa
...
-        catch ( PluginResolutionException e )
+        catch ( PluginResolutionException | InterpolationException e )
         {
             throw new ExtensionResolutionException( extension, e.getCause() );
         }
     }
 
+    private static Interpolator createInterpolatorFunction( MavenExecutionRequest request )
+    {
+        RegexBasedInterpolator interpolator = new RegexBasedInterpolator();
```
Comment (michael-o) [inline maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java:166]: Isn't the `StringSearchInterpolator` enough?
See `org.apache.maven.model.building.DefaultModelBuilder.interpolateModel(Model, ModelBuildingRequest, ModelProblemCollector)`

**Review by michael-o — CHANGES_REQUESTED** [review]
Do you want to provide default values through `jvm.config` and then override at the command line?

**Comment by laeubi** [comment]
> Do you want to provide default values through `jvm.config` and then override at the command line?

It seems how it should behave of course one can also give the value at commandline only.

**Comment by michael-o** [comment]
> > Do you want to provide default values through `jvm.config` and then override at the command line?
> 
> It seems how it should behave of course one can also give the value at commandline only.

To make sure, just try `jvm.config`, I expect it to work.

**Comment by michael-o** [comment]
Can this be accompanied with a unit or integration test?

**Comment by laeubi** [comment]
> Can this be accompanied with a unit or integration test?

I'm not very familiar with the code-base if you can point me to some tests that do already test the extensions.xml I can try to add one.

**Comment by michael-o** [comment]
> > Can this be accompanied with a unit or integration test?
> 
> I'm not very familiar with the code-base if you can point me to some tests that do already test the extensions.xml I can try to add one.

Search for `extensions.xml` [here](https://github.com/apache/maven-integration-testing/tree/master/core-it-suite/src/test/resources). I am certain that at least on IT contains this.

**Comment by laeubi** [comment]
> Search for `extensions.xml` [here](https://github.com/apache/maven-integration-testing/tree/master/core-it-suite/src/test/resources). I am certain that at least on IT contains this.

I have created https://github.com/apache/maven-integration-testing/pull/136

File: maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java (hunk 1 above)
Comment (laeubi) [reply maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java]: Done

File: maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java (hunk 2 above)
Comment (laeubi) [reply maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java]: Done.

File: maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java [hunk 3]
```diff
@@ -127,30 +134,39 @@ private CoreExtensionEntry createExtension( CoreExtension extension, List<Artifa
...
             PreorderNodeListGenerator nlg = new PreorderNodeListGenerator();
             root.accept( nlg );
 
             return nlg.getArtifacts( false );
         }
-        catch ( PluginResolutionException e )
+        catch ( PluginResolutionException | InterpolationException e )
         {
             throw new ExtensionResolutionException( extension, e.getCause() );
         }
```
[the remark this answers is not in the document]
Comment (michael-o) [reply maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java:161]: I do agree, but that should be done separately, outside of this PR.

[the remark this answers is not in the document]
Comment (laeubi) [reply maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java:161]: So should I simply add two catch clauses for this PR?

[the remark this answers is not in the document]
Comment (michael-o) [reply maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java:161]: Yes, add two please.

[the remark this answers is not in the document]
Comment (laeubi) [reply maven-embedder/src/main/java/org/apache/maven/cli/internal/BootstrapCoreExtensionManager.java:161]: Done!

**Comment by jira-importer** [comment]
Resolve #9171

**Pull request closed without merging**